feat: parse ConEmu OSC9;1

dmehala opened the pull request:
# Description

This PR implements support for the [ConEmu OSC9;1 escape sequence](https://conemu.github.io/en/AnsiEscapeCodes.html#OSC_Operating_system_commands).

Based on my understanding of [ConEmu's source code](https://github.com/Maximus5/ConEmu/blob/740b09c363cb16fbb730d72c53eaca1c530a016e/src/ConEmuCD/ConAnsiImpl.cpp#L705-L724):
- The default timeout is set to `100` milliseconds if no value is specified.  
- The timeout value is clamped to a maximum of `10000` milliseconds.

#3125

Commit:
feat: parse ConEmu OSC9;1

dmehala commented on src/terminal/osc.zig:
```diff
@@ -158,6 +158,9 @@ pub const Command = union(enum) {
     hyperlink_end: void,
 
+    /// Sleep (OSC 9;1) in ms
+    sleep: u16,
```
I considered using the [Duration type](https://github.com/ghostty-org/ghostty/blob/94599102e9fb8247af08cbbbcb7ee25e3d31e1bd/src/config/Config.zig#L5546) but this would introduce a dependency on Config. Would that be an issue?

jcollie replied:
I don't think that the Duration type is necessary here, unless there's a need to display it with units or parse it from a string units. You might consider scaling it to nanoseconds and using a `u64` since that's what Zig tends to use for that sort of thing.

mitchellh replied:
Since the duration is in the ConEmu spec as `ms` I'm fine keeping this as `u16` and documenting it (as @dmehala already has) as a millisecond unit.

dmehala commented on src/terminal/osc.zig:
```diff
@@ -788,6 +796,21 @@ pub const Parser = struct {
+            },
+
+            .conemu_sleep_value => switch (c) {
+                '0'...'9' => {},
```
I could construct the number directly here, but I thought it would be safer to handle potential overflow correctly by using `std.fmt.parseUnsigned`.

jcollie replied:
`parseUnsigned` allows `_` as in `1_000_000` is a valid Zig number so I always feel a little icky when using it to parse data. I think that doing it like this is fine. Performance shouldn't be a concern.

dmehala commented on src/terminal/osc.zig:
```diff
@@ -1147,6 +1170,22 @@ pub const Parser = struct {
+                const str = self.buf[self.buf_start..self.buf_idx];
+                if (str.len != 0) {
+                    if (std.fmt.parseUnsigned(u16, str, 10)) |num| {
+                        v.* = @min(num, 10000);
+                    } else |_| {
+                        v.* = 10000;
```
Not sure if we need to replicate ConEmu's behaviour here.

jcollie replied:
What's ConEmu's behavior? We should either replicate ConEmu's behavior or document the differences.

dmehala replied:
https://github.com/Maximus5/ConEmu/blob/740b09c363cb16fbb730d72c53eaca1c530a016e/src/ConEmuCD/ConAnsiImpl.cpp#L705-L724

jcollie replied:
OK, looks like we match ConEmu's behavior. Please be careful linking to code though as we can't use GPL code since we are MIT. ConEmu is BSD 3 clause so I think that's OK (IANAL).

jcollie commented on src/terminal/osc.zig:
```diff
@@ -1634,6 +1674,48 @@ test "OSC: set palette color" {
+    try testing.expect(cmd == .sleep);
+    try testing.expectEqual(10000, cmd.sleep);
+}
+
```
You should test invalid codes as well, perhaps parameters that contain something other than digits.

dmehala replied:
addressed in c98d207eb98b1a0d822e148c79792f1b17c3a98a

Commit:
code review

- Add test with invalid value.
- Fix inspector compilation.

dmehala commented on src/terminal/osc.zig:
```diff
@@ -1634,6 +1676,61 @@ test "OSC: set palette color" {
+    try testing.expectEqual(10000, cmd.sleep.duration_ms);
+}
+
+test "OSC: conemu sleep invalid input" {
+    const testing = std.testing;
+
+    var p: Parser = .{};
+
+    const input = "9;1;foo";
+    for (input) |ch| p.next(ch);
+
+    const cmd = p.end('\x1b');
+
+    try testing.expect(cmd == null);
+}
```
Note to myself: well that should default to `100` to fully match conemu's behaviou

dmehala replied:
addressed in 9d9fa60ece3736001be39ec2ac80b0e6c4b44f17

Commit:
code review

- Default to 100 if the value can't be parsed as an integer or
  is missing entirely.

dmehala commented:
Alright, I think the logic matches ConEmu's now.

mitchellh commented:
Please rebase. I can't merge :)

Commit:
Merge branch 'main' into dmehala/conemu-osc9
